arch/sim/src/sim/up_ucontext.c

```diff
@@ -0,0 +1,283 @@
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: task_exit_point
+ *
+ * Description:
+ *   This function is used to release the allocated stack memory for a task
+ *   on it's exit path.
+ *
+ * Assumptions:
+ *   This function does not return.
+ *
+ ****************************************************************************/
+
+static void task_exit_point(void)
+{
+  for (; ; )
+    {
+      int cpu_id = 0;
+
+#ifdef CONFIG_SMP
+      cpu_id = up_cpu_index();
+#endif
+      if (g_task_exit_details[cpu_id].ucontext_sp)
+        {
+          /* Free stack memory used by the ucontext structure */
+
+          free(g_task_exit_details[cpu_id].current_ucontext);
+          free(g_task_exit_details[cpu_id].ucontext_sp);
+
+          /* Activate the parent ucontext */
+
+          setcontext(g_task_exit_details[cpu_id].parent_ucontext);
+        }
+    }
+}
+
+/****************************************************************************
+ * Name: up_setup_exit_context
+ *
+ * Description:
+ *   Creates new exit context for tasks. In a SMP configuration every CPU
+ *   has a different exit context.
+ *
+ ****************************************************************************/
+
+static void up_setup_exit_context(int cpu)
+{
+  ucontext_t *exit_context = &g_task_exit_context[cpu];
+  int ret = getcontext(exit_context);
+  ASSERT(ret >= 0);
+
+  uint8_t *sp = &g_task_exit_context_stack[cpu][0];
+
+  uint64_t align_offset = STACK_ALIGNMENT_BYTES - ((uint64_t)sp %
+      STACK_ALIGNMENT_BYTES);
+
+  uint8_t *aligned_stack = sp + align_offset;
+  exit_context->uc_stack.ss_sp   = aligned_stack;
+  exit_context->uc_stack.ss_size = MIN_UCONTEXT_STACK_LENGTH -
+    align_offset;
+  exit_context->uc_link          = &g_uctx_main;
+
+  makecontext(exit_context, task_exit_point, 0);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_create_context
+ *
+ * Description:
+ *   Creates a new ucontext structure and initialize it.
+ *
+ * Input Parameters:
+ *   ucontext_sp   - buffer where we store the ucontext stack pointer
+ *   prev_ucontext - the parent ucontext used%i to return to when the
+ *    execution of the current context ends.
+ *   entry_point   - the entry point of the new context
+ *
+ * Return Value:
+ *   A pointer to the ucontext structure or NULL in case something went
+ *   wrong.
+ *
+ ****************************************************************************/
+
+void *up_create_context(void **ucontext_sp, void *prev_ucontext,
+    void (*entry_point)(void))
+{
+  ucontext_t *new_context = calloc(1, sizeof(ucontext_t));
```

Comment:
we can directly define a global ucontext_t array which will simplify the destroying process a lot.

Reply:
Right, we can do this and we will not need the `up_destroy_ucontext`.